horovod/common/ops/mpi_gpu_operations.cc

```diff
@@ -195,6 +199,9 @@ Status MPI_GPUAlltoall::Execute(std::vector<TensorTableEntry>& entries, const Re
   assert(entries.size() == 1);
 
   gpu_op_context_.InitGPU(entries);
+
+  WaitForData(entries);
```

Comment:
Is this `InitGPU` call why `WaitForData` needs to be in the `Execute` function instead of sitting before it? Otherwise it seems we could reduce some duplication by calling it beforehand.

Reply:
Yes, the `InitGPU` call is what lazily creates Horovod's internal CUDA stream which is required for the `WaitForData` variant that uses `cudaStreamWaitEvent`. I do also dislike the duplication but wasn't sure if there were better options. If you have any suggestions, I'm all ears.

Reply:
Thanks for clarifying, I think that's fine then. Let's land it!